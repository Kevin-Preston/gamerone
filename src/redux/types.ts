import { AuthState } from './auth/types';
import { ProfileState } from './profile/types';
import { SettingsState } from './settings/types';
import { PostState } from './post/types';
import { PostFormState } from './post-form/types';
import { RequestStatusState } from './request-status/types';
import { RouterState } from 'connected-react-router';

export const INIT_STATE = 'root/INIT_STATE';
export const LOAD_STATE = 'root/LOAD_STATE';

export interface InitStateAction {
  type: typeof INIT_STATE;
}

export type RootStateActions = InitStateAction;

export interface RootState {
  Auth: AuthState;
  Profile: ProfileState;
  Settings: SettingsState;
  Post: PostState;
  PostForm: PostFormState;
  RequestStatus: RequestStatusState;
  router: RouterState;
}
