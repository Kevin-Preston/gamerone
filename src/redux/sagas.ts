import { all } from 'redux-saga/effects';
import authSagas from './auth/saga';
import profileSagas from './profile/saga';
import settingsSagas from './settings/saga';
import postSagas from './post/saga';
import postFormSagas from './post-form/saga';
import miscSagas from './misc/saga';

export default function* rootSaga() {
  yield all([
    authSagas(),
    profileSagas(),
    settingsSagas(),
    postSagas(),
    postFormSagas(),
    miscSagas(),
  ]);
}
