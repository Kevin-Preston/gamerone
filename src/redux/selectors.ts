/**
 * Here comes the combined selectors from different states
 */

import { createSelector } from 'reselect';

import { selectCurrentUser } from './auth/selectors';

import {
  selectCurrentProfileUser,
  selectCurrentProfileNetworks,
  selectCurrentProfileSponsors,
  selectCurrentProfileProducts,
} from './profile/selectors';

import {
  selectSettingsSponsors,
  selectSettingsProducts,
  selectSettingsProfileUser,
  selectProfileSocialNetworks,
} from './settings/selectors';

// determines whether the currently viewing profile is my profile or not
export const selectIsSelfProfile = createSelector(
  [selectCurrentProfileUser, selectCurrentUser],
  (profileUser, authUser) => authUser.id === profileUser?.id,
);

export const selectProfileUser = createSelector(
  [selectCurrentProfileUser, selectSettingsProfileUser],
  (profileUser, settingsUser) =>
    settingsUser.id === profileUser?.id ? settingsUser : profileUser,
);

// resolves social networks to be shown on Profile page
export const selectSocialNetworks = createSelector(
  [
    selectIsSelfProfile,
    selectProfileSocialNetworks,
    selectCurrentProfileNetworks,
  ],
  (isSelfProfile, settingsNetworks, profileNetworks) =>
    isSelfProfile ? settingsNetworks : profileNetworks,
);

// resolves sponsors to be shown on Profile page
export const selectSponsors = createSelector(
  [selectIsSelfProfile, selectSettingsSponsors, selectCurrentProfileSponsors],
  (isSelfProfile, settingsSponsors, profileSponsors) =>
    isSelfProfile ? settingsSponsors : profileSponsors,
);

// resolves products to be shown on Profile page
export const selectProducts = createSelector(
  [selectIsSelfProfile, selectSettingsProducts, selectCurrentProfileProducts],
  (isSelfProfile, settingsProducts, profileProducts) =>
    isSelfProfile ? settingsProducts : profileProducts,
);
