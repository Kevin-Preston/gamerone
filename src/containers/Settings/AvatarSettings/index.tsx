import React, { useState, useEffect } from 'react';
import { Dispatch, AnyAction } from 'redux';
import { connect } from 'react-redux';
import { useDropzone } from 'react-dropzone';
import Avatar from 'components/common/Avatar';
import Button, { ButtonSchemeEnum } from 'components/common/Button';
import Card from 'components/common/Card';
import EditImageCard from 'containers/Cards/EditImageCard';
import { selectCurrentUser } from 'redux/auth/selectors';
import { selectUploadAvatarStatus } from 'redux/request-status/selectors';
import SettingsActions from 'redux/settings/actions';
import RequestStatusActions from 'redux/request-status/actions';
import { AVATAR_WIDTH, AVATAR_HEIGHT } from 'utils/constants';
import { RootState } from 'redux/types';
import { UPLOAD_AVATAR_REQUEST } from 'redux/settings/types';

import './style.scss';

interface Props {
  submitTitle?: string;
}
const AvatarForm: React.FC<Props & AvatarFormProps> = ({
  user,
  status,
  submitTitle = 'Upload',
  dispatchUploadAvatar,
  dispatchCleanForm,
}: Props & AvatarFormProps): JSX.Element => {
  const [editing, setEditing] = useState(false);
  const { username, avatar } = user;
  const [uploadedAvatar, setUploadedAvatar] = useState<string | null>(
    user.avatar,
  );
  const { getRootProps, getInputProps } = useDropzone({
    accept: 'image/*',
    multiple: false,
    onDrop: (acceptedFiles) => {
      const files = acceptedFiles.map((file) =>
        Object.assign(file, {
          preview: URL.createObjectURL(file),
        }),
      );
      setEditing(true);
      setUploadedAvatar(files[0].preview);
    },
  });

  useEffect(() => {
    setUploadedAvatar(user.avatar);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [user]);

  useEffect(() => {
    return () => {
      dispatchCleanForm();
    };
  }, [dispatchCleanForm]);

  const handleCancel = () => {
    setUploadedAvatar(avatar);
    setEditing(false);
  };

  const handleApply = (blobUrl: string) => {
    setUploadedAvatar(blobUrl);
    setEditing(false);
  };

  const handleUpload = async (e: any) => {
    e.preventDefault();
    if (!uploadedAvatar) return;

    const blob = await fetch(uploadedAvatar).then((r: any) => r.blob());
    const file = new File([blob], 'f.png', {
      type: blob.type,
      lastModified: Date.now(),
    });

    dispatchUploadAvatar(file);
  };

  return (
    <>
      <Card>
        <form>
          <div className="card__content">
            <h4>Profile Image</h4>

            {status?.isError && (
              <div role="alert">Something went wrong. Please try again</div>
            )}

            <div {...getRootProps({ className: 'pointer' })}>
              <input {...getInputProps()} aria-label="file-upload" />
              <Avatar
                src={uploadedAvatar ? uploadedAvatar : avatar}
                alt={username || ''}
                size={'xl'}
                className="avatar-upload"
              />
            </div>
          </div>
          <div className="card__actions">
            <Button
              scheme={ButtonSchemeEnum.PRIMARY}
              onClick={handleUpload}
              submitting={status?.isFetching}
              disabled={avatar === uploadedAvatar}
            >
              {submitTitle}
            </Button>
          </div>
        </form>
      </Card>
      {uploadedAvatar && (
        <EditImageCard
          image={uploadedAvatar}
          visible={editing}
          title="Edit Profile Image"
          onCancel={handleCancel}
          onApply={handleApply}
          width={AVATAR_WIDTH / 2}
          height={AVATAR_HEIGHT / 2}
          borderRadius={100}
        />
      )}
    </>
  );
};

const mapStateToProps = (state: RootState) => {
  return {
    user: selectCurrentUser(state),
    status: selectUploadAvatarStatus(state),
  };
};

export function mapDispatchToProps(dispatch: Dispatch<AnyAction>) {
  return {
    dispatchUploadAvatar: (f: File) =>
      dispatch(SettingsActions.uploadAvatar(f)),
    dispatchCleanForm: () =>
      dispatch(RequestStatusActions.cleanStatus(UPLOAD_AVATAR_REQUEST)),
  };
}

type AvatarFormProps = ReturnType<typeof mapStateToProps> &
  ReturnType<typeof mapDispatchToProps>;

export const ConnectedAvatarForm = connect(
  mapStateToProps,
  mapDispatchToProps,
)(AvatarForm);

export default function AvatarSettings() {
  return <ConnectedAvatarForm />;
}
