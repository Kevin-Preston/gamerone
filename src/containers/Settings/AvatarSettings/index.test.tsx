import React from 'react';
import { Provider } from 'react-redux';
import { createProduct } from 'redux';
import { render } from '@testing-library/react';
import { act } from 'react-dom/test-utils';
import authReducer from 'redux/auth/reducer';
import AvatarSettings from './index';

const testUser = {
  id: 1449,
  username: 'vadim',
  email: 'vadimas.workteam@gmail.com',
  firstName: 'frontend',
  lastName: 'dev',
  bio: null,
  avatar: 'profile/vadim.svg',
  banner: null,
  playCount: 0,
  gameCount: 0,
  badgeCount: 0,
  followCount: 0,
  followerCount: 0,
  birthDate: null,
};

const renderWithAuthReducer = (
  component = <AvatarSettings />,
  {
    initialState,
    store = createProduct(authReducer, { Auth: initialState }),
  } = {},
) => {
  let utils = {};
  act(() => {
    utils = render(<Provider store={store}>{component}</Provider>);
  });

  return utils;
};

describe('avatar settings', () => {
  test('renders the form without crash', async () => {
    const { asFragment } = renderWithAuthReducer(<AvatarSettings />, {
      initialState: { user: testUser },
    });

    expect(asFragment()).toMatchSnapshot();
  });

  test('shows edit image modal on image load', async () => {
    renderWithAuthReducer(<AvatarSettings />, {
      initialState: { user: testUser },
    });

    // TODO: Simulate file load event

    // global.URL.createObjectURL = jest.fn(() => 'details');
    // window.navigator.msSaveOrOpenBlob = jest.fn(() => 'details');

    // const file = new File(['(⌐□_□)'], 'avatar.png', { type: 'image/png' });
    // const fileElm = getByLabelText(/file-upload/);

    // act(() => {
    //   fireEvent.change(fileElm, { target: { files: [file] } });
    // });
  });

  test('disable the form button on load', () => {
    const { getByRole } = renderWithAuthReducer(<AvatarSettings />, {
      initialState: { user: testUser },
    });

    expect(getByRole('button', { name: 'Upload' }).disabled).toEqual(true);
  });

  test('renders the user avatar on load', () => {
    const { getByRole } = renderWithAuthReducer(<AvatarSettings />, {
      initialState: { user: testUser },
    });

    expect(
      getByRole('img', { name: testUser.username }).src.includes(
        testUser.avatar,
      ),
    ).toBe(true);
  });
});
