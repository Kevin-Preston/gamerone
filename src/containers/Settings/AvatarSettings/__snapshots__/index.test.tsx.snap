// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`avatar settings renders the form without crash 1`] = `
<DocumentFragment>
  <div
    class="card"
  >
    <form>
      <div
        class="card__content"
      >
        <h4>
          Avatar
        </h4>
        <div
          class="dropzone form-group text-center"
          tabindex="0"
        >
          <input
            accept="image/*"
            aria-label="file-upload"
            autocomplete="off"
            style="display: none;"
            tabindex="-1"
            type="file"
          />
          <img
            alt="vadim"
            class="image_avatar xl avatar-upload"
            height="128"
            id="avatar"
            src="https://cdn.gamerone.gg/128x128/profile/vadim.svg"
            width="128"
          />
        </div>
      </div>
      <div
        class="card__actions"
      >
        <button
          class="button button--primary"
          disabled=""
          type="submit"
        >
          Upload
        </button>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
